Improved PE blocks and code

gentlegiantJGC opened the pull request:
_readSubChunk_1plus was getting the entity and tile entity data once for each subchunk then passing it to add_data where it was only being used for the first subchunk. Split add_data into add_terrain and add_data. add_terrain only deals with the terrain and is called 16 times per chunk and add_data only deals with entities and tile entities and is called once per chunk

cleaned up version variable names to make it more obvious which variables hold which values
chunk_version - the version found for that chunk in the leveldb
subchunk_version - the encoding version for that subchunk
world_version - already defined but cleaned up

gentlegiantJGC committed:
cleaned up versions and split chunk.add_data

_readSubChunk_1plus was getting the entity and tile entity data once for each subchunk then passing it to add_data where it was only being used for the first subchunk. Split add_data into add_terrain and add_data. add_terrain only deals with the terrain and is called 16 times per chunk and add_data only deals with entities and tile entities and is called once per chunk

cleaned up version variable names to make it more obvious which variables hold which values
chunk_version - the version found for that chunk in the leveldb
subchunk_version - the encoding version for that subchunk
world_version - already defined but cleaned up

gentlegiantJGC committed:
changed idStr to match the new string ids and added banners

The new bedrock chunk loading code looks for the entry with the same string id but a lot were using the old string id or a wrong string id (this might have been my fault as I didn't understand what the entry was for)

gentlegiantJGC committed:
fixed sub-chunks not saving if they didn't already exist

Previously for data to be written to disk the chunk and sub-chunk had to exist when MCedit was loaded. I wrote an MCedit filter to create them but thought I would look into fixing this internally so the filter was not needed.
The sub-chunk key only needs to exist in the database if it contains blocks that are not just air so the key is deleted if there is only air.

gentlegiantJGC committed:
Further sub-chunk fix

Apparently what the previous commit fixed had already been fixed however it was saving air chunks when they do not need saving.
Improved upon the previous commit with this knowledge to make it more efficient

naor2013 commented on pymclevel/leveldbpocket.py:
```diff
@@ -348,17 +348,7 @@ def _readSubChunk_1plus(self, cx, cz, y, readOptions=None, key=None):
-                entities = None
-
-        return terrain, tile_entities, entities
+        return terrain
```
Why did you remove that part and made it outside? It makes more sense that the "read_chunk" function will give you all the details and that it won't be outside

naor2013 commented on pymclevel/leveldbpocket.py:
```diff
@@ -414,25 +404,34 @@ def _readChunk(self, cx, cz, world, readOptions=None):
+                    entities = db.Get(rop, key + "\x32")
+                except RuntimeError:
+                    entities = None
+                chunk.add_data(tile_entities=tile_entities, entities=entities)
```
I think it makes more sense to have one "add_data" and not separate them to the different types of data

naor2013 commented on pymclevel/leveldbpocket.py:
```diff
@@ -497,42 +496,59 @@ def _saveChunk_1plus(self, chunk, batch=None, writeOptions=None):
                 skyLight = packNibbleArray(chunk._SkyLight.binary_data[y]).tostring()
                 blockLight = packNibbleArray(chunk._BlockLight.binary_data[y]).tostring()
-            terrain = ver + blocks + blockData + skyLight + blockLight
+            if blocks is not None and blockData is not None:
```
I think "or" instead of "and" will make more sense... Just in case

naor2013 replied:
Actually, that is fine, I misread the line

naor2013 commented on pymclevel/leveldbpocket.py:
```diff
@@ -497,42 +496,59 @@ def _saveChunk_1plus(self, chunk, batch=None, writeOptions=None):
             chunk._BlockLight.update_subchunks()
         chunk.subchunks = chunk._Blocks.subchunks
 
+        empty = numpy.zeros((16, 16, 16), dtype="uint8")
```
Makes sense to put it down where it's used

naor2013 reviewed (changes requested):
I added some comments... Love this pull request in general, just a few small things to fix

gentlegiantJGC committed:
re-merged splitting of add_data

add_data was previously split to solve an issue where data was read more than needed. This has now been merged again and the issue solved in a different way to keep the add_data function as a single function

gentlegiantJGC committed:
removed try: except:

the code reviewer was complaining about this and it doesn't even seem like it is needed

gentlegiantJGC committed:
improving checking if empty

noar suggested moving this down and I realised there was a much more efficient way of doing it
